### PR TITLE
Allow IamPolicy to read json policies where the principal value itself an array

### PR DESCRIPTION
**Pull request by cenedhryn**
## Motivation and Context
The [IamPolicy](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/iam-policy-builder/src/main/java/software/amazon/awssdk/policybuilder/iam/IamPolicy.java) class represents an IAM policy. You can serialize/deserialize this class to json. A principal is represented as a type/value in [IamPrincipal](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/iam-policy-builder/src/main/java/software/amazon/awssdk/policybuilder/iam/IamPrincipal.java). 

If you have a principal of for instance "AWS" and it contains more than one value, you can create it by doing
`IamPrincipal.createAll("AWS", asList("arn1", "arn2"))`, which yields a list of `IamPrincipal` with the same type and the two different values. When reading from json, the principal is correctly interpreted if supplied in this format. 

However a valid representation of principals in the console for instance is as "AWS" : ["arn1", "arn1"]. This format cannot be parsed into a valid IamPolicy. This PR allows the IamPolicy to parse compound principals. Note they will still be represented in the list format above

## Modifications
Use `IamPrincipal.createAll()` which can handle duplicates.

## Testing
Unit tests

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services-custom/iam-policy-builder/src/main/java/software/amazon/awssdk/policybuilder/iam/internal/DefaultIamPolicyReader.java [hunk 1]
```diff
@@ -132,9 +132,10 @@ private List<IamPrincipal> readPrincipals(Map<String, JsonNode> statementObject,
 
         if (principalsNode.isObject()) {
             List<IamPrincipal> result = new ArrayList<>();
-            principalsNode.asObject().forEach((id, value) -> {
-                result.add(IamPrincipal.create(id, expectString(value, name + " entry value")));
-            });
+            Map<String, JsonNode> principalsNodeObject = principalsNode.asObject();
+            principalsNodeObject.keySet().forEach(
+                k -> result.addAll(IamPrincipal.createAll(k, readStringArray(principalsNodeObject, k)))
+            );
```
Comment (dagnir) [inline services-custom/iam-policy-builder/src/main/java/software/amazon/awssdk/policybuilder/iam/internal/DefaultIamPolicyReader.java:138]: Does this still work for non-array principals?

Like

```
"Principal": { "AWS": "123456789012" }
```

Comment (ccr-amz) [reply services-custom/iam-policy-builder/src/main/java/software/amazon/awssdk/policybuilder/iam/internal/DefaultIamPolicyReader.java:138]: [readStringArray](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/iam-policy-builder/src/main/java/software/amazon/awssdk/policybuilder/iam/internal/DefaultIamPolicyReader.java#L180) method is capable of reading non-array principals as well.

Comment (dagnir) [reply services-custom/iam-policy-builder/src/main/java/software/amazon/awssdk/policybuilder/iam/internal/DefaultIamPolicyReader.java:138]: Cool (if not confusing)! @cenedhryn Can we rename this method while we're here? Maybe read**As**StringArray?

Comment (cenedhryn) [reply services-custom/iam-policy-builder/src/main/java/software/amazon/awssdk/policybuilder/iam/internal/DefaultIamPolicyReader.java]: I went super literal!

**Review by dagnir — APPROVED** [review]

**Pull request merged**